genie-web/src/main/java/com/netflix/genie/web/agent/services/impl/AgentConfigurationServiceImpl.java

```diff
@@ -0,0 +1,138 @@
+import java.util.regex.Pattern;
+
+/**
+ * This implementation of {@link AgentConfigurationService} forwards properties set on the server that match a given
+ * set of regular expressions, plus any additional ones specified in configuration.
+ * It utilizes a cache to avoid recomputing the set of properties and values for every request.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+public class AgentConfigurationServiceImpl implements AgentConfigurationService {
+
+    private static final String AGENT_PROPERTIES_CACHE_KEY = "agent-properties";
+
+    private final AgentConfigurationProperties agentConfigurationProperties;
+    private final Environment environment;
+    private final Pattern agentPropertiesPattern;
+    private final LoadingCache<String, Map<String, String>> cache;
+
+    /**
+     * Constructor.
+     *
+     * @param agentConfigurationProperties the properties
+     * @param environment the environment
+     */
+    public AgentConfigurationServiceImpl(
+        final AgentConfigurationProperties agentConfigurationProperties,
+        final Environment environment
+    ) {
+        this.agentConfigurationProperties = agentConfigurationProperties;
+        this.environment = environment;
+
+        this.agentPropertiesPattern = Pattern.compile(
+            this.agentConfigurationProperties.getAgentPropertiesFilterPattern(),
+            Pattern.CASE_INSENSITIVE
+        );
+
+        // Use a cache to re-compute agent properties periodically, rather than for every request.
+        this.cache = Caffeine
+            .newBuilder()
+            .expireAfterWrite(this.agentConfigurationProperties.getCacheExpirationInterval())
+            .refreshAfterWrite(this.agentConfigurationProperties.getCacheRefreshInterval())
+            .initialCapacity(1)
+            .build(this::loadProperties);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Map<String, String> getAgentProperties() {
+        return this.cache.get(AGENT_PROPERTIES_CACHE_KEY);
+    }
+
+
+    private Map<String, String> loadProperties(@NonNull final String propertiesKey) {
+        if (AGENT_PROPERTIES_CACHE_KEY.equals(propertiesKey)) {
+            return reloadAgentProperties();
+        }
+        throw new IllegalArgumentException("Unknown key to load: " + propertiesKey);
+    }
+
+    private Map<String, String> reloadAgentProperties() {
+        // Obtain properties sources from environment
+        final PropertySources propertySources;
+        if (environment instanceof ConfigurableEnvironment) {
+            propertySources = ((ConfigurableEnvironment) environment).getPropertySources();
+        } else {
+            propertySources = new StandardEnvironment().getPropertySources();
+        }
+
+        // Create set of all properties that match the filter pattern
+        final Set<String> filteredPropertyNames = Sets.newHashSet();
+        for (final PropertySource<?> propertySource : propertySources) {
```

Comment:
Is this the only way to flatten and list all the properties? I'd have hoped somewhere in the environment there was a better API

Reply:
I looked at Spring's `EnvironmentEndpoint` to see how they "list all properties" and this is what they do (in their case it's a lot more complex because they want to keep them separate and annotated with origin, etc). But same basic approach.